File: pi-daemon-rust/src/serial_integrity.rs
use crc32fast::Hasher;

const SERIAL_INTEGRITY_MAGIC: u16 = 0x8AC1;
pub const SERIAL_INTEGRITY_HDR_LEN: usize = 8;

pub fn wrap_payload(payload: &[u8], out: &mut Vec<u8>) {
    let mut hasher = Hasher::new();
    hasher.update(payload);
    let crc = hasher.finalize();

    out.clear();
    out.extend_from_slice(&SERIAL_INTEGRITY_MAGIC.to_be_bytes());
    out.extend_from_slice(&(payload.len() as u16).to_be_bytes());
    out.extend_from_slice(&crc.to_be_bytes());
    out.extend_from_slice(payload);
}

pub fn unwrap_payload(frame: &[u8]) -> Result<&[u8], String> {
    if frame.len() < SERIAL_INTEGRITY_HDR_LEN {
        return Err("Frame too short for integrity header".into());
    }

    let magic = u16::from_be_bytes([frame[0], frame[1]]);
    if magic != SERIAL_INTEGRITY_MAGIC {
        return Err(format!("Invalid magic: {:#06x}", magic));
    }

    let length = u16::from_be_bytes([frame[2], frame[3]]) as usize;
    if length != frame.len() - SERIAL_INTEGRITY_HDR_LEN {
        return Err(format!("Length mismatch: expected {}, got {}", length, frame.len() - SERIAL_INTEGRITY_HDR_LEN));
    }

    let crc = u32::from_be_bytes([frame[4], frame[5], frame[6], frame[7]]);
    let payload = &frame[SERIAL_INTEGRITY_HDR_LEN..];

    let mut hasher = Hasher::new();
    hasher.update(payload);
    let computed_crc = hasher.finalize();

    if computed_crc != crc {
        return Err(format!("CRC mismatch: expected {:#010x}, got {:#010x}", crc, computed_crc));
    }

    Ok(payload)
}

File: pi-daemon/serial_integrity.py
import struct
import zlib

# [ Magic (2B) | PayloadLen (2B) | PayloadCRC32 (4B) | Payload... ]
# Magic: 0x8A 0xC1
SERIAL_INTEGRITY_MAGIC = 0x8AC1
SERIAL_INTEGRITY_HDR_LEN = 8
SERIAL_INTEGRITY_MAX_PAYLOAD_LEN = 1280
_HEADER_STRUCT = struct.Struct(">HHI")

def wrap_payload(payload: bytes) -> bytes:
    if len(payload) > SERIAL_INTEGRITY_MAX_PAYLOAD_LEN:
        raise ValueError(
            f"Payload too large: {len(payload)} > {SERIAL_INTEGRITY_MAX_PAYLOAD_LEN}"
        )
    crc = zlib.crc32(payload) & 0xFFFFFFFF
    header = _HEADER_STRUCT.pack(SERIAL_INTEGRITY_MAGIC, len(payload), crc)
    return header + payload

def unwrap_payload(frame: bytes) -> bytes:
    if len(frame) < SERIAL_INTEGRITY_HDR_LEN:
        raise ValueError("Frame too short for integrity header")
    magic, length, crc = _HEADER_STRUCT.unpack(frame[:SERIAL_INTEGRITY_HDR_LEN])
    if magic != SERIAL_INTEGRITY_MAGIC:
        raise ValueError(f"Invalid magic: {hex(magic)}")
    if length != len(frame) - SERIAL_INTEGRITY_HDR_LEN:
        raise ValueError(f"Length mismatch: expected {length}, got {len(frame) - SERIAL_INTEGRITY_HDR_LEN}")
    if length > SERIAL_INTEGRITY_MAX_PAYLOAD_LEN:
        raise ValueError(
            f"Payload too large: {length} > {SERIAL_INTEGRITY_MAX_PAYLOAD_LEN}"
        )
    payload = frame[SERIAL_INTEGRITY_HDR_LEN:]
    computed_crc = zlib.crc32(payload) & 0xFFFFFFFF
    if computed_crc != crc:
        raise ValueError(f"CRC mismatch: expected {hex(crc)}, got {hex(computed_crc)}")
    return payload
